setup.py

```diff
@@ -395,6 +335,7 @@ def finalize_options(self):
         'scipy >= 0.18.1',
         'six >= 1.5.0',
         'smart_open >= 1.8.1',
+        'cython',
```

Comment:
Why is this necessary?

If a user installs from a binary wheels, they don't need Cython for anything.

And if they install from source, they should be using the precompiled sources (`.c`), so they still don't need Cython?

Reply:
Yes, they don't need it, but it's not difficult to install, so there is no harm in including it.

What is the benefit of including *.c sources, when they can be regenerated from Cython anyway?

Reply:
Distributing generated `.c` is the official recommended way:
https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#distributing-cython-modules

But I actually don't know why. We could easily pin the Cython version, to ensure repeatable cythonization. Is Cython really so painless to install? CC @menshikh-iv @gojomo do you remember why we do it this way?

Btw if users install from a binary wheel, these dependencies would still kick in, right? We definitely don't want to force users to install a compiler (Windows etc).

Reply:
First, to answer your questions: yeah, with wheels, the binaries will already be in place, so Cython or compilation will happen.

Reading through your link, I can see two reasons why distributing .c and .cpp is the recommended way:

1. The installed Cython version may be incompatible with the source, and fail to compile correctly
2. Cython may not be installed all together, forcing the user to install it
3. Even if the correct Cython version is installed, the user "probably does not want to use it just to compile your module"

They don't seem to be particularly compelling. 1) and 2) are kind of weak, because installing Cython seems trivial (`pip install cython==1.2.3`) if we decide to pin the Cython version. 3) just sounds like bullshit to me: if people are building things from source, it's reasonable for them to expect to use a compiler at some stage.

That doesn't leave much justification for the "strong recommendation", in my opinion. Perhaps there are other reasons we are not aware of?

My main motivations for **not** including the .c and .cpp sources are:

1. Much simpler setup.py
2. Less maintenance work because we don't have to keep the sources in version control

@menshikh-iv WDYT?

Reply:
@piskvorky 

> is Cython really so painless to install?

From my experience - yes, I never get this kind of problem.


>  do you remember why we do it this way?

No, this was done before I came to gensim, hope than @gojomo remember.


@mpenkov 

I'm +1 with pinning Cython version (to avoid any potential incompatibilities).

> if people are building things from source, it's reasonable for them to expect to use a compiler at some stage.


I agree, sometimes this really needed (in case if you want to make a contribution).
I also supports both of your points (setup.py & less maintenance). I see no practical reasons to store "auto-generated" files in a git repo.

Reply:
> First, to answer your questions: yeah, with wheels, the binaries will already be in place, so Cython or compilation will happen.

You mean "will **not** happen", right?

If that's the case = Windows users can continue to enjoy Gensim without having to install compilers, then I'm +1 on requiring Cython (=compiler) if installing from source (as opposed to binary wheel).

I'm thinking that maybe we had to make Cython optional before because of the pure-Python code paths. But that's no longer an issue.

Reply:
Yes, sorry, I misspoke. Compilation will **not** happen.

Reply:
`cython` has not previously been a declared prerequisite, and was only necessary for developers who were modifying the `.pyx` (etc) sources. Before wheels, every install required a local compilation of the `.c` files. That setup predates any of my contributions. (Although the wheels clearly help on Windows, I'm not sure they're actually a better experience everywhere else. I liked the clarity of having all source that was possibly running on your machine locally; the wheels introduce the risk that a corrupted/subverted build process sends users binaries that don't match public source.) Personally I wouldn't add `cython` as a declared requirement - why trigger installers to install something, with the storage costs and even executable-code risks, that isn't really needed?

Reply:
I've amended setup.py to only install cython when it is needed (e.g. if the C/C++ files are missing).